README.md
# Belgium's Energy Consumption Visualization (Group B)

### Deployed App Links

Chad Wheeler: [https://py-prototype.herokuapp.com/](https://py-prototype.herokuapp.com/)

Harpreet Kaur: [https://powerful-spire-82147.herokuapp.com/](https://powerful-spire-82147.herokuapp.com/)

Nelson Tang: [https://data551-milestone2.herokuapp.com/](https://data551-milestone2.herokuapp.com/)

Nyanda Redwood: [energy-app](https://energy-app.herokuapp.com/)

## Who are we

- Chad: "An intelligent and witty person that works on the betterment of oneself and those around him." -Jacob
- Harpreet: Passionate mathematician on the path of becoming a statistician.
- Nelson: Definitely a data engineer on my right track in the near future.
- Nyanda: Life long learner vested with a growth mentality.

## What are we doing and Why

We built a dashboard to visualize data related to a low energy-house in Belgium. Our purpose is to provide analytical insights, visually, for ease of use in cases such as the determination of the adequate photovoltaic and energy storage size. Ideally, this could be of utmost importance to Belgium's energy management system. This is particularly true in the context of diminishing the inflow of power to the grid across the four seasons. This is premised on avoiding power balckouts. On the ground, this could be of immediate use to electricity grid operators in their calculations as they strategize how to avoid power blackouts.

What is unique about our proposal is the use of a low-energy household to generate the data. This usage could be valuable in deciding upon a minimum threshold for energy consumption per household. This can be of further use in managing load control within unforseen emergency situations as well as in the creation of backup systems.  

Our app allows users, such as an operator's supervising team, to flexibly filter their search and view different aspects of the data by filtering and re-ordering on different variables.

## How can you get involved

If you are interested in analyzing the variation of temperature and energy consumption for parts of the winter season(January to February) or parts of the the spring season(March to April), this dashboard can be of help.

With this dashboard, you recieve an interactive layout of various charts that represent several different types of information related to energy consumption. Using a dropdown menu, you can choose from a compilation of nine rooms and/or from a compilation of four months, separately.

This dashboard provides you with concise reports based on day(s), week(s), and month(s). This supports analytical interests related to energy management systems within the context of a low-energy house in Belgium. For example, the vizualizations offer the granularity for comparitive insights based on weekend energy consumption versus weekday energy consumption.




<img src ="docs/layout2.png" width="500px">

## Insight into data

The data set was donated in the year 2017 by Luis Candanedo, University of Mons (UMONS). It was collected to create regression models of appliances energy use in a low energy building in Belgium. Our dataset includes approximately, 20000 temperature and humidity sensors measurements from a ZigBee wireless network, outside weather conditions (`To` and `RH_out`, `Wind speed`, `Visibility`, `Tdewpoint`) from a nearby airport station (Chievres Airport, Belgium), and recordings of the energy consumed by the lighting fixtures and other appliances operated by the people residing in that house. The recordings have been recorded at 10 min for about 4.5 months (i.e. from January 11, 2016 to May 27, 2016) in a low-energy building. The node transmitted the temperature recording (`T1` to `T9`) in Celsius and humidity (`RH_1` to `RH_9`) in % for every individual room in the house around every 3.3 minutes which were then averaged for 10 minutes periods. The energy data for both appliances (`Appliances`) and lights (`lights`) separately was logged every 10 minutes with m-bus energy meters in the Wh unit.

## How to get started with this dashboard

This dashboard offers dynamic date range selection, allowing you to input dates according to your interest. Date range selections can go from a single day to span an entire month or even several months. Each date range selection simultaneously updates all plots; there are five plots in total.

For the total energy consumption of all appliance, you can view the bar chart within the dashboard. This displays Watts per hour for the house according to your date range selection. Adjancent to the bar chart is a pie chart, which demonstrates a proportional comparision of the filtered selection.

For a month-wise comparison of energy consumption by appliance and light, you can view the area chart labeled "Energy Used in House".

If you are interested in knowing about room temperature and humidity, there is a drop-down menu selection that allows you to do this. This selection updates the line chart labeled "Temperature and Humidity Trend".

If you are interested in knowing the temperature and humidity outside the house, you can view the area chart labled " Outside Temperature vs Humidity", which displays the humdity in percentage and the temperature in Celsius.

## Acknowledgements and references

- <http://archive.ics.uci.edu/ml/datasets/Appliances+energy+prediction>
- <https://www.sciencedirect.com/science/article/abs/pii/S0378778816308970?via%3Dihub>


changes.txt
add opacity to the plot
make nyanada`s plot not as area but as line
changes in the additional file

> add limitation of colour scheme sync

src/TimeFramePlots.py
import pandas as pd
from datetime import datetime, date
from dash import Dash, html, dcc
from dash.dependencies import Input, Output
import altair as alt

# Read in global data
energydata = pd.read_csv("../data/energydata_complete.csv")

# create a day of week column and month column and day column
energydata['date'] = pd.to_datetime(energydata['date'])
energydata['day_of_week'] = energydata["date"].dt.day_name()
energydata['month'] = energydata["date"].dt.strftime('%b')
energydata['day'] = energydata["date"].dt.date

app = Dash(__name__, external_stylesheets=['https://codepen.io/chriddyp/pen/bWLwgP.css'])


def bar_plot_altair(start_date, end_date):
    # select date range
    selected_data = energydata[(energydata['day'] <= end_date) & (energydata['day'] >= start_date)]
    # group by day of week and sum the energy consumption
    selected_data = selected_data.groupby(['day_of_week']).sum().reset_index()

    # make Appliances vs Day of week bar plot
    chart = alt.Chart(selected_data).mark_bar(color='#D35400', size=20).encode(
        x=alt.X('Appliances:Q', axis=alt.Axis(title='Appliances energy consumption in Wh')),
        y=alt.Y('day_of_week:N', axis=alt.Axis(title='Day of the week'),
                sort=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday'])
    ).properties(
        width=400,
        height=300
    )
    return chart.to_html()


def pie_chart(start_date, end_date):
    # select date range
    selected_data = energydata[(energydata['day'] <= end_date) & (energydata['day'] >= start_date)]
    # group by day of week and sum the energy consumption
    selected_data = selected_data.groupby(['day_of_week']).sum().reset_index()

    chart = alt.Chart(selected_data).mark_arc(innerRadius=50).encode(
        theta=alt.Theta(field="Appliances", type="quantitative"),
        color=alt.Color(field="day_of_week", type="nominal", legend=alt.Legend(title="Day of the week"),
                        sort=['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']),
    )
    return chart.to_html()


app.layout = html.Div([

    dcc.DatePickerRange(
        id='my-date-picker-range',
        min_date_allowed=date(2016, 1, 11),
        max_date_allowed=date(2016, 5, 27),
        initial_visible_month=date(2016, 1, 11),
        start_date=date(2016, 1, 11),
        end_date=date(2016, 5, 27)
    ),

    html.Iframe(
        id='barchart',
        srcDoc=bar_plot_altair(start_date=energydata['day'].min(), end_date=energydata['day'].max()),
        style={'border-width': '0', 'width': '100%', 'height': '400px'}),

    html.Iframe(
        id='piechart',
        srcDoc=pie_chart(start_date=energydata['day'].min(), end_date=energydata['day'].max()),
        style={'border-width': '0', 'width': '100%', 'height': '400px'})
])


@app.callback(
    Output('barchart', 'srcDoc'),
    Output('piechart', 'srcDoc'),
    Input('my-date-picker-range', 'start_date'),
    Input('my-date-picker-range', 'end_date'))
def update_output(start_date, end_date):
    # convert the outputs to date object
    start_date_object = date.fromisoformat(start_date)
    end_date_object = date.fromisoformat(end_date)
    return (bar_plot_altair(start_date=start_date_object, end_date=end_date_object),
            pie_chart(start_date=start_date_object, end_date=end_date_object))


if __name__ == '__main__':
    app.run_server(debug=True)


src/app.py
import dash
import altair as alt
import pandas as pd
import dash_bootstrap_components as dbc
import pathlib
from dash import html
from dash import dcc
from dash.dependencies import Input, Output
from datetime import date

alt.data_transformers.disable_max_rows()

root_dir = pathlib.Path(__file__).parent.parent
file_path = root_dir.joinpath("data/energydata_complete.csv")
energydata = pd.read_csv(file_path)

# create a day of week column and month column and day column
energydata["date"] = pd.to_datetime(energydata["date"])
energydata["day_of_week"] = energydata["date"].dt.day_name()
energydata["month"] = energydata["date"].dt.strftime("%b")
energydata["day"] = energydata["date"].dt.date
energydata["day"] = pd.to_datetime(energydata["day"])

sort_order = ["Jan", "Feb", "Mar", "Apr", "May"]


app = dash.Dash(__name__, external_stylesheets=[dbc.themes.BOOTSTRAP])
server = app.server


## Plotting
# bar chart for day of week
def bar_plot_altair(start_date, end_date):
    # select date range

    """
    Presents the plot for day_of_week and sums of the energy consumped separately for all the day of that week. 

    Parameters
    ----------
    df : pd.DataFrame(selected_data)
        Groups the initial 
        energydata_complete data by day_of_week.
    start_date : start date selected
    end_date : end date selected

    Returns
    -------
    The summarized bar-plot.

    """

    selected_data = energydata[
        (energydata["day"] <= pd.to_datetime(end_date))
        & (energydata["day"] >= pd.to_datetime(start_date))
    ]
    # group by day of week and sum the energy consumption
    selected_data = selected_data.groupby(["day_of_week"]).sum().reset_index()

    # make Appliances vs Day of week bar plot
    chart = (
        alt.Chart(selected_data)
        .mark_bar(color="#D35400", size=20)
        .encode(
            x=alt.X(
                "Appliances:Q",
                axis=alt.Axis(title="Appliances energy consumption in Wh"),
            ),
            y=alt.Y(
                "day_of_week:N",
                axis=alt.Axis(title="Day of the week"),
                sort=[
                    "Monday",
                    "Tuesday",
                    "Wednesday",
                    "Thursday",
                    "Friday",
                    "Saturday",
                    "Sunday",
                ],
            ),
        )
        .properties(width=300, height=300, title="Energy Used in Each Day of Week")
    )
    return chart.to_html()


def plot_outsidetemp(start_date, end_date, xcol="T_out"):
    # select date range

    """
    Presents the trend describing a comparitive study of the temperature and humidity outside.

    Parameters
    ----------
    df : pd.DataFrame
        Uses initial energydata_complete data.
    start_date : start date selected
    end_date : end date selected

    Returns
    -------
    The summarized area plot.

    """

    selected_data = energydata[
        (energydata["day"] <= pd.to_datetime(end_date))
        & (energydata["day"] >= pd.to_datetime(start_date))
    ]
    chart = (
        alt.Chart(selected_data)
        .mark_area()
        .encode(
            x=alt.X(
                xcol,
                axis=alt.Axis(title="Temperature Outside in Celsius"),
            ),
            y=alt.X(
                "RH_out",
                axis=alt.Axis(title="Humidity outside in %"),
            ),
        )
        .properties(width=350, height=300, title="Outside Temperature vs Humidity")
    )

    return chart.interactive().to_html()


# pie chart for day of week
def pie_chart(start_date, end_date):
    # select date range

    """
    Presents the day of the week proportion summary of the energy consumed by the appliances in the house.

    Parameters
    ----------
    df : pd.DataFrame(selected_data)
        Groups the initial dataframe by day_of_week.
    start_date : start date selected
    end_date : end date selected

    Returns
    -------
    The summarized pie chart.

    """

    selected_data = energydata[
        (energydata["day"] <= pd.to_datetime(end_date))
        & (energydata["day"] >= pd.to_datetime(start_date))
    ]
    # group by day of week and sum the energy consumption
    selected_data = selected_data.groupby(["day_of_week"]).sum().reset_index()

    chart = (
        alt.Chart(selected_data)
        .mark_arc(innerRadius=50)
        .encode(
            theta=alt.Theta(field="Appliances", type="quantitative"),
            color=alt.Color(
                field="day_of_week",
                type="nominal",
                legend=alt.Legend(title="Day of the week"),
                sort=[
                    "Monday",
                    "Tuesday",
                    "Wednesday",
                    "Thursday",
                    "Friday",
                    "Saturday",
                    "Sunday",
                ],
            ),
        )
        .configure_axis(grid=False)
        .configure_view(strokeWidth=0)
        .properties(width=250, height=300)
    )
    return chart.to_html()


def plot_temp_hum(start_date, end_date, room="T1RH_1"):
    # select date range

    """
    Presents the temperature and humidity comparision in relation to the date.

    Parameters
    ----------
    df : pd.DataFrame(selected_data)
        Uses initial energydata_complete data.
    start_date : start date selected
    end_date : end date selected

    Returns
    -------
    Presents the trend charts for the temperature and humidity.

    """

    selected_data = energydata[
        (energydata["day"] <= pd.to_datetime(end_date))
        & (energydata["day"] >= pd.to_datetime(start_date))
    ]

    room_temp = room[0:2]
    room_hum = room[2:6]

    temp = (
        alt.Chart(selected_data)
        .mark_line(color="red")
        .encode(
            alt.Y(room_temp + ":Q", title="Temperature"),
            alt.X("date", title="Date"),
        )
    )

    hum = (
        alt.Chart(selected_data)
        .mark_line(color="blue")
        .encode(
            alt.Y(room_hum + ":Q", title="Humidity"),
            alt.X("date", title="Date"),
        )
    )

    plot = (
        alt.layer(temp, hum)
        .resolve_scale(y="independent")
        .properties(width=320, height=300, title="Temperature and Humidity Trend")
    )

    return plot.to_html()


def area_plot(start_date, end_date):

    """
    Presents the comparitive study of the light and appliances energy consumed in relation to individual month.

    Parameters
    ----------
    df : pd.DataFrame(selected_data)
        Subset of the initial energydata_complete data by month name.
    start_date : start date selected
    end_date : end date selected

    Returns
    -------
    Presents the area chart for the energy consumed by lights and appliances.

    """    

    selected_data = energydata[
        (energydata["day"] <= pd.to_datetime(end_date))
        & (energydata["day"] >= pd.to_datetime(start_date))
    ]

    energy_data_subset = selected_data[["Appliances", "lights", "month"]]

    energy_data_subset = (
        energy_data_subset.groupby("month", sort=False).sum().reset_index()
    )
    energy_data_subset = pd.melt(
        energy_data_subset, id_vars=["month"], value_vars=["Appliances", "lights"]
    )

    plot = (
        alt.Chart(energy_data_subset)
        .mark_area()
        .encode(
            x=alt.X(
                "month",
                sort=sort_order,
                axis=alt.Axis(
                    title="Month",
                    tickCount=10,
                    grid=False,
                    labelAngle=-360,
                    titleFontSize=12,
                    labelFontSize=10,
                ),
                scale=alt.Scale(zero=False, domain=list(sort_order)),
            ),
            y=alt.Y(
                "value",
                axis=alt.Axis(
                    title="Energy use in Wh",
                    grid=False,
                    titleFontSize=12,
                    labelFontSize=10,
                ),
                scale=alt.Scale(zero=False),
            ),
            color=alt.Color("variable"),
        )
        .properties(height=300, width=350, title="Energy Used in House")
        .configure_axis(labelFontSize=14, titleFontSize=18)
        .configure_mark(opacity=0.5)
    )

    return plot.to_html()

# line plot for temperature and humidity for day
temp_hum = html.Iframe(
    id="temp_hum",
    srcDoc=plot_temp_hum(
        start_date=energydata["day"].min(), end_date=energydata["day"].max()
    ),
    style={"width": "100%", "height": "400px"},
)

# bar chart for day of week
energy_bar = html.Iframe(
    id="barchart",
    srcDoc=bar_plot_altair(
        start_date=energydata["day"].min(), end_date=energydata["day"].max()
    ),
    style={"width": "100%", "height": "400px"},
)

# pie chart for day of week
energy_pie = html.Iframe(
    id="pie_chart",
    srcDoc=pie_chart(
        start_date=energydata["day"].min(), end_date=energydata["day"].max()
    ),
    style={"width": "100%", "height": "400px"},
)

# line plot for temperature inside and outside for day
temp_hum_out = html.Iframe(
    id="altair_chart",
    srcDoc=plot_outsidetemp(
        start_date=energydata["day"].min(),
        end_date=energydata["day"].max(),
        xcol="T_out",
    ),
    style={"width": "100%", "height": "400px"},
)

# area plot for appliances and energy consumption for day
energy_month = html.Iframe(
    id="energy_month",
    srcDoc=area_plot(
        start_date=energydata["day"].min(), end_date=energydata["day"].max()
    ),
    style={"width": "130%", "height": "400px"},
)

header = dcc.Markdown(
    """
        # Energy Use of Appliance in a Low-Energy House        
    """
)

random_text = dcc.Markdown(
    """ 
        **Collaborators:**  
        Harpreet Kaur  
        Chad Wheeler  
        Nelson Tang  
        Nyanda Redwood   
    """
)

date_range = dcc.Markdown(
    """
        **Select date range:** 
        """
)

pick_room = dcc.Markdown(
    """
        &nbsp
        
        **Pick a room:**  
        
        
    """
)


# dropdown to select room
room_dropdown = dcc.Dropdown(
    id="room",
    value="T1RH_1",
    options=[
        {"label": "Room 1", "value": "T1RH_1"},
        {"label": "Room 2", "value": "T2RH_2"},
        {"label": "Room 3", "value": "T3RH_3"},
        {"label": "Room 4", "value": "T4RH_4"},
        {"label": "Room 5", "value": "T5RH_5"},
        {"label": "Room 6", "value": "T6RH_6"},
        {"label": "Room 7", "value": "T7RH_7"},
        {"label": "Room 8", "value": "T8RH_8"},
        {"label": "Room 9", "value": "T9RH_9"},
    ],
    clearable=False,
)

# date range controller
date_picker = dcc.DatePickerRange(
    id="my-date-picker-range",
    min_date_allowed=date(2016, 1, 11),
    max_date_allowed=date(2016, 5, 27),
    initial_visible_month=date(2016, 1, 11),
    start_date=date(2016, 1, 11),
    end_date=date(2016, 5, 27),
)

# layout 
row = html.Div(
    [
        dbc.Row(
            [
                dbc.Col(
                    html.Div(
                        [random_text, date_range, date_picker, pick_room, room_dropdown]
                    )
                ),
                dbc.Col(html.Div(temp_hum)),
                dbc.Col(html.Div(temp_hum_out)),
            ]
        ),
        dbc.Row(
            [
                dbc.Col(html.Div(energy_bar)),
                dbc.Col(html.Div(energy_pie)),
                dbc.Col(html.Div(energy_month)),
            ]
        ),
    ]
)

app.layout = dbc.Container(
    [
        header,
        html.Div(
            [
                row,
            ]
        ),
    ]
)


@app.callback(
    Output("temp_hum", "srcDoc"),
    Output("barchart", "srcDoc"),
    Output("pie_chart", "srcDoc"),
    Output("altair_chart", "srcDoc"),
    Output("energy_month", "srcDoc"),
    Input("room", "value"),
    Input("my-date-picker-range", "start_date"),
    Input("my-date-picker-range", "end_date"),
)
def update_output(room, start_date, end_date):
    # convert the outputs to date object

    """
    Presents the output by modifying it as per users selection

    Parameters
    ----------
    start_date : start date selected
    end_date : end date selected
    room : room selected

    Returns
    -------
    Presents the area chart for the energy consumed by lights and appliances.

    """ 

    start_date_object = date.fromisoformat(start_date)
    end_date_object = date.fromisoformat(end_date)
    return (
        plot_temp_hum(
            start_date=start_date_object, end_date=end_date_object, room=room
        ),
        bar_plot_altair(start_date=start_date_object, end_date=end_date_object),
        pie_chart(start_date=start_date_object, end_date=end_date_object),
        plot_outsidetemp(
            start_date=start_date_object, end_date=end_date_object, xcol="T_out"
        ),
        area_plot(start_date=start_date_object, end_date=end_date_object),
    )


if __name__ == "__main__":
    app.run_server(debug=True)